documentation/intermediate-ca.md

```diff
@@ -0,0 +1,191 @@
+
+    puppet config set --section main certificate_revocation leaf
+
+See [PUP-3788](https://tickets.puppetlabs.com/browse/PUP-3788) for more information on chain revocation checking.
+
+### Generate an intermediate CA certificate request
+
+    #!/opt/puppetlabs/puppet/bin/ruby
+    require 'puppet'
+    require 'openssl'
+
+    class Puppet::Application::GenCACSR < Puppet::Application
+      run_mode :master
+
+      def main
+        root_cert = OpenSSL::X509::Certificate.new(File.read(Puppet[:cacert]))
+        ca_key = OpenSSL::PKey::RSA.new(File.read(Puppet[:cakey]))
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        puts csr.to_s
+      end
+    end
+    Puppet::Application::GenCACSR.new.run
+
+Use this CSR to generate the intermediate CA certificate that Puppet will be using.
+
+### Construct the CA certificate chain
+
+Puppet Server needs to present the full certificate chain to clients in order for the client to authenticate the server. The certificate chain is constructed by concatenating the CA certificates together, starting with the new intermediate CA certificate and descending to the root CA certificate.
+
+    cat puppet-intermediate-certificate.pem \
+        org-intermediate-certificate.pem \
+         org-root-certificate.pem > ca-bundle.pem
+
+### Verify the Puppet master host certificate against the CA bundle
+
+Before installing the new certificate bundle, verify that the bundle can be used to verify the existing host certificate on the CA server. If this step fails then the CA certificate bundle is invalid and cannot be installed.
+
+    openssl verify -CAfile ca-bundle.pem $(puppet master --configprint hostcert)
+
+### Update the Puppet CA certificate bundle
+
+    cp ca-bundle.pem /etc/puppetlabs/puppet/ssl/ca/ca_crt.pem
+
+### Update the Puppet agent CA certificate bundle
+
+The Puppet agent caches the CA certificate and will not automatically fetch the new CA certificate bundle from the master. This caveat applies to the Puppet CA server itself as well. On the CA server, copy the CA certificate bundle to the local CA certificate path:
+
+    cp /etc/puppetlabs/puppet/ssl/ca/ca_crt.pem /etc/puppetlabs/puppet/ssl/certs/ca.pem
+
+If any Puppet agents have already been created, copy the CA certificate bundle to `/etc/puppetlabs/puppet/ssl/certs/ca.pem`.
```

Comment:
Can you clarify that this file needs to be copied to this directory on every agent? I think they might think we mean locally on the master.